### PR TITLE
feat: add path option to `toMatchSnapshot`

### LINKED ISSUE #7792: [Feature]: Can I somehow change the directory for saving screenshots?
### Your question

I used this https://playwright.dev/java/docs/test-snapshots function
Can I somehow change the directory for saving screenshots?

### PR DESCRIPTION
**Pull request by nickofthyme**
allow `snapshotName` to accept an array of path segments in `toMathchSnapshot`.

```ts
// example.spec.ts
import { test, expect } from '@playwright/test';

test('example test', async ({ page }) => {
  await page.goto('https://playwright.dev');
  expect(await page.screenshot()).toMatchSnapshot(['test', 'path', 'landing.png']);
});
```

This would set and lookup the screenshot from `example.spec.ts-snapshots/test/path/landing-<suffix>.png`.

> The path segment is contained to within the base snapshot directory for the given test file (e.g. `example.spec.ts-snapshots/`). Any path resolving to outside of this directory is ignored.

The current functionality of `snapshotName` as a `'string'` is preserved. This string will continue to be sanitized for file path as is currently done.

```ts
// example.spec.ts
import { test, expect } from '@playwright/test';

test('example test', async ({ page }) => {
  await page.goto('https://playwright.dev');
  expect(await page.screenshot()).toMatchSnapshot('test/path/landing.png');
});
```
This would set and lookup the screenshot from `example.spec.ts-snapshots/test-path-landing-<suffix>.png`.

Related to #7792

### CODE REVIEW

File: src/test/util.ts [hunk 1]
```diff
@@ -144,4 +144,14 @@ export function sanitizeForFilePath(s: string) {
...
 
+/**
+ * prevent parent directory traversal
+ */
+export function sanitizeForParentPath(parentPath: string, p?: string): string {
+  if (!p) return '';
+
+  const joinedPath = path.join(parentPath, path.normalize(p));
+  return /\.{2,}\//.test(path.relative(parentPath, joinedPath)) ? '' : path.normalize(p);
+}
```
Comment (nickofthyme) [inline src/test/util.ts:155]: I am open to better ways of doing this.

File: src/test/util.ts [hunk 2]
```diff
@@ -144,4 +144,14 @@ export function sanitizeForFilePath(s: string) {
...
+
```
Comment (nickofthyme) [inline src/test/util.ts:156]: I am open to better ways of doing this.

File: tests/playwright-test/golden.spec.ts [hunk 3]
```diff
@@ -488,6 +488,54 @@ test('should write missing expectations with sanitized snapshot name', async ({r
...
+  expect(result.exitCode).toBe(0);
+});
+
+test('should set snapshot with path', async ({ runInlineTest }, testInfo) => {
+  const result = await runInlineTest({
+    ...files,
+    'a.spec.js': `
+      const { test } = require('./helper');
+      test('is a test', ({}) => {
+        expect('Hello world').toMatchSnapshot('snapshot.txt', { path: 'test/path' });
```
Comment (dgozman) [inline tests/playwright-test/golden.spec.ts:512]: Should we instead partially revert #7620? The problem we were facing there was `path` going outside of the snapshots directory.

If we make `toMatchSnapshot('test/path/snapshot.txt')` to place snapshot at `a.spec.js-snapshots/test/path/snapshot.txt` that should just work, shouldn't it? This will save on extra params and APIs.

File: src/test/util.ts (hunk 2 above)
Comment (dgozman) [reply src/test/util.ts:156]: Perhaps we should just do `path.resolve()` instead and then check that `resultingPath.startsWith(baseSnapshotDir + path.sep)`?

**Review by dgozman — COMMENTED** [review]
Thank you for the PR! See comments inline.

File: tests/playwright-test/golden.spec.ts (hunk 3 above)
Comment (nickofthyme) [reply tests/playwright-test/golden.spec.ts]: Yeah good call I didn't notice that was a recent change. See changes in 9b9b677.

File: src/test/util.ts (hunk 2 above)
Comment (nickofthyme) [reply src/test/util.ts]: Yeah I like that approach better. See change in 0d0510a.

**Comment by nickofthyme** [comment]
@dgozman I'm curious what your thoughts are on adding a `config` option much like [`outputDir`](https://playwright.dev/docs/api/class-testconfig#test-config-output-dir) for `snapshotDir` that would be the base directory for all snapshots seeing as this would the main request for #7792. For example...

```ts
// playwright.config.ts
import { PlaywrightTestConfig } from '@playwright/test';

const config: PlaywrightTestConfig = {
  snapshotDir: 'snapshots',
};

export default config;
```

```
playwright.config.ts
tests/
└── foo.spec.ts
snapshots/
└── foo.spec.ts-snapshots/
     ├── basic-test-Chrome-linux.png
     └── basic-test-Safari-linux.png
```

Would there be any issue with this? I have started on these changes but think I'll put them in a follow up PR to this one.

File: src/test/util.ts [hunk 4]
```diff
@@ -144,4 +144,13 @@ export function sanitizeForFilePath(s: string) {
   return s.replace(/[\x00-\x2F\x3A-\x40\x5B-\x60\x7B-\x7F]+/g, '-');
 }
 
+/**
+ * Get absolute path contained within parent directory, otherwise returns parent path
+ */
+export function getContainedPath(parentPath: string, subPath?: string): string {
+  if (!subPath) return parentPath;
+  const resolvedPath = path.resolve(parentPath, subPath);
+  return resolvedPath.startsWith(parentPath) ? resolvedPath : path.join(parentPath, path.basename(subPath));
```
Comment (dgozman) [inline src/test/util.ts:153]: I think this needs to be `startsWith(parentPath  + path.sep)`, otherwise `toMatchSnapshot('../foo.spec.js-snapshots123/bar.txt')` will escape.

File: src/test/util.ts [hunk 5]
```diff
@@ -144,4 +144,13 @@ export function sanitizeForFilePath(s: string) {
   return s.replace(/[\x00-\x2F\x3A-\x40\x5B-\x60\x7B-\x7F]+/g, '-');
 }
 
+/**
+ * Get absolute path contained within parent directory, otherwise returns parent path
```
Comment (dgozman) [inline src/test/util.ts:148]: Why don't we throw a nice error instead of constructing an unexpected path? I think that would be appreciated by the user.

File: src/test/workerRunner.ts [hunk 6]
```diff
@@ -242,20 +242,23 @@ export class WorkerRunner extends EventEmitter {
       outputDir: baseOutputDir,
       outputPath: (...pathSegments: string[]): string => {
         fs.mkdirSync(baseOutputDir, { recursive: true });
-        return path.join(baseOutputDir, ...pathSegments);
+        return getContainedPath(baseOutputDir, path.join(...pathSegments));
       },
       snapshotPath: (snapshotName: string): string => {
```
Comment (dgozman) [inline src/test/workerRunner.ts:247]: Sorry for not spotting it earlier, but perhaps we want `...pathSegments: string[]` here, similar to `outputPath`? This way we can make it strictly-backwards-compatible (if we bring back the `sanitizeForFilePath()` call for each path segment), but also allow for the new usecase.

File: docs/src/test-snapshots-js.md [hunk 7]
```diff
@@ -55,7 +55,8 @@ Sometimes you need to update the reference screenshot, for example when the page
 npx playwright test --update-snapshots
 ```
 
-Note that `snapshotName` is *not a path* relative to the test file, so don't try to use it like `expect(value).toMatchSnapshot('../../test-snapshots/snapshot.png')`.
+> Note that `snapshotName` also accepts a relative path to the snapshot file such as `expect(value).toMatchSnapshot('test-snapshots/snapshot.png')`.
+> However, this path is enfored to be within the snapshots directory for each test file (i.e. `a.spec.js-snapshots`) and will be ignored otherwise.
```
Comment (dgozman) [inline docs/src/test-snapshots-js.md:59]: - typo: enfored -> enforced
- I'd make it throw when it goes outside of the snapshots directory, and so change this to `However, this path must stay within the snapshots directory for each test file (i.e. a.spec.js-snapshots), otherwise it will throw.`
- A similar comment in `class-testinfo.md` for `outputPath` and `snapshotPath` methods would be great!

**Review by dgozman — COMMENTED** [review]
I am sorry that this PR turns into a lot of work, but it is going to be a great feature!

File: src/test/workerRunner.ts (hunk 6 above)
Comment (nickofthyme) [reply src/test/workerRunner.ts:247]: This sound good to keep BC. Just to clarify your idea. So `snapshotPath` now takes `pathSegments` like `outputPath` and sanitizes each of them before joining. But `outputPath` will **not** sanitize the path segments which could make the two paths inconsistent. I realize they are independent directories but they mirror each other currently, Is this acceptable?

Also what is your thoughts about the `toMatchSnapshot(name)` api? Would it remain with `name` and then we would just need to do `snapshotPath(name.split(path.sep))`? Or do you have ideas on maintaining this BC as well?

Comment (dgozman) [reply src/test/workerRunner.ts:247]: Great questions!

I did confuse `snapshotPath()` and `toMatchSnapshot()` apis. The latter one is important, and we cannot make `...pathSegments` there. But we can go with `toMatchSnapshot(name: string | string[], options?)` and let the user pass an array explicitly. That's not as convenient, but seems like a niche feature. Also, passing an array instead of `foo/bar/baz` makes it cross-platform out of the box because you don't need to worry about path separators. WDYT?

Re: different behavior of `outputPath` vs `snapshotPath` - I think that's fine. We can also move sanitization into `toMatchSnapshot` and only apply it when you pass a single string (as before), and avoid sanitization when given an array. This way, `TestInfo` methods will behave similarly, while `toMatchSnapshot` will be backwards compatible. Does this make sense?

File: docs/src/test-snapshots-js.md (hunk 7 above)
Comment (nickofthyme) [reply docs/src/test-snapshots-js.md]: Good ideas, fixed in e899857.

BTW do I need to manually update `docs/src/test-api/class-testinfo.md` and `docs/src/test-snapshots-js.md` from `types/test.d.ts` or is there some auto generation script I am missing?

File: src/test/util.ts (hunk 5 above)
Comment (nickofthyme) [reply src/test/util.ts]: Wasn't sure what the protocol was for errors but I like the idea. Let me know what you think of df1e578

File: src/test/util.ts (hunk 4 above)
Comment (nickofthyme) [reply src/test/util.ts]: I don't exactly follow, are saying in the event someone passes a path the goes to the parent then back into the correct snapshot directory it will not escape the path? If I were to test something like...

```ts
test.only('should prevent reassigning snapshot test path', async ({ runInlineTest }) => {
  const result = await runInlineTest({
    ...files,
    'a.spec.js-snapshots/snapshot.txt': `Hello world`,
    'a.spec.js': `
      const { test } = require('./helper');
      test('is a test', ({}) => {
        expect('Hello world').toMatchSnapshot(['..', 'a.spec.js-snapshots', 'snapshot.txt']);
      });
    `
  });

  expect(result.exitCode).toBe(0);
});
```

This test fails because `outputPath` path is expecting `/test-results/a-is-a-test` and not `/test-results/a.spec.js-snapshots` since the `outputPath` is formatted differently than snapshots.

```
  1) a.spec.js:6:7 › is a test =====================================================================

    Error: Path defined outside of parent directory
      path: ../a.spec.js-snapshots/snapshot.txt
      resolved path: /Users/nicholaspartridge/Documents/repos/playwright/test-results/golden-should-prevent-reassigning-snapshot-test-path-playwright-test/test-results/a.spec.js-snapshots/snapshot.txt
      parent: /Users/nicholaspartridge/Documents/repos/playwright/test-results/golden-should-prevent-reassigning-snapshot-test-path-playwright-test/test-results/a-is-a-test

      5 |       const { test } = require('./helper');
      6 |       test('is a test', ({}) => {
    > 7 |         expect('Hello world').toMatchSnapshot(['..', 'a.spec.js-snapshots', 'snapshot.txt']);
        |                               ^
      8 |       });
```

This seems like an edge case that could be resolved with the new error message. Is your intent for this case to throw on `snapshotPath`?

File: docs/src/test-snapshots-js.md [hunk 8]
```diff
@@ -55,7 +55,8 @@ Sometimes you need to update the reference screenshot, for example when the page
 npx playwright test --update-snapshots
 ```
 
-Note that `snapshotName` is *not a path* relative to the test file, so don't try to use it like `expect(value).toMatchSnapshot('../../test-snapshots/snapshot.png')`.
+> Note that `snapshotName` also accepts a relative path to the snapshot file such as `expect(value).toMatchSnapshot('test-snapshots/snapshot.png')`.
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply docs/src/test-snapshots-js.md]: Fixed in e899857

File: src/test/workerRunner.ts (hunk 6 above)
Comment (nickofthyme) [reply src/test/workerRunner.ts]: Ok yeah that clears up my concerns, I think that all makes sense. Here are my changes based on my understanding 93d9632. Let me know if I missed something. Happy to update anything big or small 👍🏼.

**Comment by nickofthyme** [comment]
I wasn't sure about the naming in the docs for `toMatchSnapshot` and changing it from `snapshotName` to `pathSegments`. Might be best to keep it as is with `snapshotName` with the note about using it as `pathSegments`. Also looks like I might need to fix some tests once you take another look.

File: tests/playwright-test/golden.spec.ts [hunk 9]
```diff
@@ -472,22 +471,91 @@ test('should sanitize snapshot name', async ({ runInlineTest }) => {
...
+  }, { 'update-snapshots': true });
+
+  expect(result.exitCode).toBe(0);
+  const snapshotOutputPath = testInfo.outputPath('a.spec.js-snapshots/test/path/snapshot.txt');
   expect(result.output).toContain(`${snapshotOutputPath} is missing in snapshots, writing actual`);
   const data = fs.readFileSync(snapshotOutputPath);
   expect(data.toString()).toBe('Hello world');
 });
 
+test.only('should attach expected/actual/diff with snapshot path', async ({ runInlineTest }, testInfo) => {
```
Comment (dgozman) [inline tests/playwright-test/golden.spec.ts:520]: Don't forget to remove `test.only` 😄

File: tests/playwright-test/test-output-dir.spec.ts [hunk 10]
```diff
@@ -216,6 +216,40 @@ test('should include the project name', async ({ runInlineTest }) => {
...
+      });
+      test('test with parent path', async ({}, testInfo) => {
+        console.log(testInfo.snapshotPath('../test/path/bar.txt').replace(/\\\\/g, '/'));
+      });
+    `,
+  });
+
+  expect(result.exitCode).toBe(0);
+  expect(result.results[0].status).toBe('passed');
+  expect(result.results[1].status).toBe('passed'); // reverts to base snapshot path
```
Comment (dgozman) [inline tests/playwright-test/test-output-dir.spec.ts:247]: So, this one should throw now, right?

**Review by dgozman — COMMENTED** [review]
I think this works nicely, just a few code comments.

> BTW do I need to manually update docs/src/test-api/class-testinfo.md and docs/src/test-snapshots-js.md from types/test.d.ts or is there some auto generation script I am missing?

Files in `docs/src` are the source of truth, `types/test.d.ts` is generated from them by `npm run watch`.

> I wasn't sure about the naming in the docs for toMatchSnapshot and changing it from snapshotName to pathSegments. Might be best to keep it as is with snapshotName with the note about using it as pathSegments

I agree that it's better to keep `snapshotName`.

File: tests/playwright-test/golden.spec.ts [hunk 11]
```diff
@@ -456,12 +456,11 @@ test('should respect project threshold', async ({ runInlineTest }) => {
   expect(result.exitCode).toBe(0);
 });
 
-test('should sanitize snapshot name', async ({ runInlineTest }) => {
+test('should sanitize snapshot name when passed as string', async ({ runInlineTest }) => {
   const result = await runInlineTest({
-    ...files,
-    'a.spec.js-snapshots/-snapshot-.txt': `Hello world`,
+    'a.spec.js-snapshots/-snapshot--darwin.txt': `Hello world`,
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply tests/playwright-test/golden.spec.ts:461]: Thanks I was wondering where that `-darwin` came from. These spec files took me a second to figure out.

File: src/test/matchers/toMatchSnapshot.ts [hunk 12]
```diff
@@ -40,10 +42,11 @@ export function toMatchSnapshot(this: ReturnType<Expect['getState']>, received:
   if (options.threshold === undefined && projectThreshold !== undefined)
     options.threshold = projectThreshold;
 
+  const pathSegments = Array.isArray(options.name) ? options.name : [addSuffix(options.name, '', undefined, true)];
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/matchers/toMatchSnapshot.ts:45]: Yeah I was trying to reuse the suffix code, but the sanitize makes it extra fancy 😄. Happy to put this back inline.

File: src/test/matchers/golden.ts [hunk 13]
```diff
@@ -82,15 +83,15 @@ function compareText(actual: Buffer | string, expectedBuffer: Buffer): { diff?:
...
+  pathSegments: string[],
+  snapshotPath: TestInfoImpl['snapshotPath'],
+  outputPath: TestInfoImpl['outputPath'],
   updateSnapshots: UpdateSnapshots,
   withNegateComparison: boolean,
   options?: { threshold?: number }
 ): { pass: boolean; message?: string; expectedPath?: string, actualPath?: string, diffPath?: string, mimeType?: string } {
-  const snapshotFile = snapshotPath(name);
-  const outputFile = outputPath(name);
+  const snapshotFile = snapshotPath(...pathSegments);
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/matchers/golden.ts:93]: Ah gotcha I thought you wanted it the same as `outputPath` 👍🏼

File: tests/playwright-test/test-output-dir.spec.ts (hunk 10 above)
Comment (nickofthyme) [reply tests/playwright-test/test-output-dir.spec.ts]: Yes correct, 311fa94.

File: tests/playwright-test/golden.spec.ts (hunk 9 above)
Comment (nickofthyme) [reply tests/playwright-test/golden.spec.ts]: Whoops 😅  311fa94

File: tests/playwright-test/golden.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply tests/playwright-test/golden.spec.ts]: 311fa94

File: src/test/workerRunner.ts [hunk 14]
```diff
@@ -255,20 +255,20 @@ export class WorkerRunner extends EventEmitter {
...
         if (testInfo.snapshotSuffix)
           suffix += '-' + testInfo.snapshotSuffix;
-        const ext = path.extname(snapshotName);
-        if (ext)
-          snapshotName = sanitizeForFilePath(snapshotName.substring(0, snapshotName.length - ext.length)) + suffix + ext;
-        else
-          snapshotName = sanitizeForFilePath(snapshotName) + suffix;
-        return path.join(test._requireFile + '-snapshots', snapshotName);
+
+        const snapshotPath = path.join(...pathSegments.map((segment, i, { length }) =>
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/workerRunner.ts]: 311fa94

File: src/test/matchers/toMatchSnapshot.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/matchers/toMatchSnapshot.ts]: 311fa94

File: src/test/util.ts [hunk 15]
```diff
@@ -144,4 +144,26 @@ export function sanitizeForFilePath(s: string) {
   return s.replace(/[\x00-\x2F\x3A-\x40\x5B-\x60\x7B-\x7F]+/g, '-');
 }
 
+export function addSuffix(filePath: string, suffix: string, customExtension?: string, sanitize = false): string {
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/util.ts]: 311fa94

File: src/test/matchers/golden.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/matchers/golden.ts]: I changed this and array in 311fa94 but lmk if you actually wanted the spread.

File: src/test/util.ts [hunk 16]
```diff
@@ -144,4 +144,26 @@ export function sanitizeForFilePath(s: string) {
...
+
+/**
+ * Returns absolute path contained within parent directory.
+ */
+export function getContainedPath(parentPath: string, subPath: string = ''): string {
+  const resolvedPath = path.resolve(parentPath, subPath);
+
+  if (resolvedPath.startsWith(parentPath)) return resolvedPath;
+
+  throw new Error(`Path defined outside of parent directory
```
[the remark this answers is not in the document]
Comment (nickofthyme) [reply src/test/util.ts]: That sounds good to me. I wasn't sure what the best way to do this so I just returned `null` if outside of the parent and throw an error at both callsites. See 07567bc. Let me know if you have any better ideas for doing this or the error message.

File: tests/playwright-test/golden.spec.ts [hunk 17]
```diff
@@ -472,22 +472,91 @@ test('should sanitize snapshot name', async ({ runInlineTest }) => {
...
+      const { test } = pwt;
       test('is a test', ({}) => {
         expect('Hello world').toMatchSnapshot('../../snapshot!.txt');
       });
     `
   }, {}, { CI: '' });
 
   expect(result.exitCode).toBe(1);
-  const snapshotOutputPath = testInfo.outputPath('a.spec.js-snapshots/-snapshot-.txt');
+  const snapshotOutputPath = testInfo.outputPath('a.spec.js-snapshots/-snapshot--darwin.txt');
```
Comment (dgozman) [inline tests/playwright-test/golden.spec.ts:484]: Use `helper` and drop `darwin` here.

File: tests/playwright-test/golden.spec.ts [hunk 18]
```diff
@@ -472,22 +472,91 @@ test('should sanitize snapshot name', async ({ runInlineTest }) => {
...
+  expect(result.output).toContain(`${snapshotOutputPath} is missing in snapshots, writing actual`);
+  const data = fs.readFileSync(snapshotOutputPath);
+  expect(data.toString()).toBe('Hello world');
+});
+
+test('should join array of snapshot path segments without sanitizing ', async ({ runInlineTest }) => {
+  const result = await runInlineTest({
+    'a.spec.js-snapshots/test/path/snapshot-darwin.txt': `Hello world`,
```
Comment (dgozman) [inline tests/playwright-test/golden.spec.ts:492]: Same here.

File: tests/playwright-test/test-output-dir.spec.ts [hunk 19]
```diff
@@ -216,6 +216,61 @@ test('should include the project name', async ({ runInlineTest }) => {
...
+    'my-test.spec.js': `
+      const { test } = require('./helper');
+      test('test with parent path', async ({}, testInfo) => {
+        console.log(testInfo.snapshotPath(['..', 'test', 'path', 'bar.txt']).replace(/\\\\/g, '/'));
+      });
+    `,
+  });
+
+  expect(result.exitCode).toBe(1);
+  expect(result.results[0].status).toBe('failed');
```
Comment (dgozman) [inline tests/playwright-test/test-output-dir.spec.ts:271]: `expect(result.output).toContain('<insert the error message you throw in the method>')`

File: docs/src/test-advanced-js.md [hunk 20]
```diff
@@ -120,7 +120,7 @@ In addition to everything from the [`workerInfo`](#workerinfo), the following in
 - `timeout: number` - Test timeout.
 - `annotations` - [Annotations](./test-annotations.md) that were added to the test.
 - `snapshotSuffix: string` - Suffix used to locate snapshots for the test.
-- `snapshotPath(snapshotName: string)` - Function that returns the full path to a particular snapshot for the test.
+- `snapshotPath(pathSegments: string[])` - Function that returns the full path to a particular snapshot for the test.
```
Comment (dgozman) [inline docs/src/test-advanced-js.md:123]: > I changed this and array in 311fa94 but lmk if you actually wanted the spread.

Sorry, I was confused by some lint error. Spread is backwards compatible, so spread for sure! Luckily (or actually not!) there are still some tests failing, so this is not the only change to be made 😕

**Review by dgozman — COMMENTED** [review]
Apart from the spread and some failing tests, this looks perfect!

File: docs/src/test-advanced-js.md (hunk 20 above)
Comment (nickofthyme) [reply docs/src/test-advanced-js.md]: No worries, simple fix. See 76efbc2.

File: docs/src/test-advanced-js.md [hunk 21]
```diff
@@ -120,7 +120,7 @@ In addition to everything from the [`workerInfo`](#workerinfo), the following in
 - `timeout: number` - Test timeout.
 - `annotations` - [Annotations](./test-annotations.md) that were added to the test.
 - `snapshotSuffix: string` - Suffix used to locate snapshots for the test.
-- `snapshotPath(snapshotName: string)` - Function that returns the full path to a particular snapshot for the test.
+- `snapshotPath(...pathSegments: string[])` - Function that returns the full path to a particular snapshot for the test.
```
Comment (nickofthyme) [inline docs/src/test-advanced-js.md:123]: This seems to control argument types for `TestInfo` inside `types/test.d.ts` but for some reason it keeps overriding the type for `snapshotPath` in `types/test.s.ts` to `snapshotPath: (pathSegments: string[]) => string;`. Any ideas?

File: tests/playwright-test/test-output-dir.spec.ts (hunk 19 above)
Comment (nickofthyme) [reply tests/playwright-test/test-output-dir.spec.ts:271]: 76efbc2

File: tests/playwright-test/golden.spec.ts (hunk 18 above)
Comment (nickofthyme) [reply tests/playwright-test/golden.spec.ts]: 76efbc2

File: tests/playwright-test/golden.spec.ts (hunk 17 above)
Comment (nickofthyme) [reply tests/playwright-test/golden.spec.ts]: 76efbc2

File: docs/src/test-advanced-js.md (hunk 21 above)
Comment (nickofthyme) [reply docs/src/test-advanced-js.md:123]: Ugh nvm it was `utils/generate_types/overrides-test.d.ts` ac578b1

File: src/test/util.ts [hunk 22]
```diff
@@ -144,4 +144,23 @@ export function sanitizeForFilePath(s: string) {
...
+  return (sanitize ? sanitizeForFilePath(base) : base) + suffix + (customExtension || ext);
+}
+
+/**
+ * Returns absolute path contained within parent directory.
+ */
+export function getContainedPath(parentPath: string, subPath: string = ''): string | null {
+  const resolvedPath = path.resolve(parentPath, subPath);
+
+  if (resolvedPath.startsWith(parentPath + path.sep)) return resolvedPath;
```
Comment (nickofthyme) [inline src/test/util.ts:161]: Can you explain why `path.sep` is needed here? it's causing a test failure for an empty `''` path here https://github.com/microsoft/playwright/pull/9156/checks?check_run_id=3762066712.

Comment (dgozman) [reply src/test/util.ts:161]: That's to avoid `getContainedPath('/foo/bar', '/foo/barbaz')` returning `true`. I guess this can be fixed with
```js
if (resolvedPath === parentPath || resolvedPath.startsWith(parentPath + path.sep)) return resolvedPath;
```

Comment (nickofthyme) [reply src/test/util.ts]: 9bc8041

File: tests/page/page-request-fulfill.spec.ts [hunk 23]
```diff
@@ -217,7 +217,7 @@ it('should fulfill with fetch result', async ({ page, server, isElectron }) => {
   expect(await response.json()).toEqual({ 'foo': 'bar' });
 });
 
-it('should fulfill with fetch result and overrides', async ({ page, server, isElectron }) => {
+it.only('should fulfill with fetch result and overrides', async ({ page, server, isElectron }) => {
```
Comment (dgozman) [inline tests/page/page-request-fulfill.spec.ts:220]:
```suggestion
it('should fulfill with fetch result and overrides', async ({ page, server, isElectron }) => {
```

File: tests/playwright-test/test-output-dir.spec.ts [hunk 24]
```diff
@@ -167,7 +167,7 @@ test('should include the project name', async ({ runInlineTest }) => {
       });
       test2('test 2', async ({}, testInfo) => {
         console.log(testInfo.outputPath('bar.txt').replace(/\\\\/g, '/'));
-        console.log(testInfo.snapshotPath('bar.txt').replace(/\\\\/g, '/'));
+        console.log(testInfo.snapshotPath('bar.txt').replace(/\\\\/g, ''));
```
Comment (dgozman) [inline tests/playwright-test/test-output-dir.spec.ts:170]:
```suggestion
        console.log(testInfo.snapshotPath('bar.txt').replace(/\\\\/g, '/'));
```

**Review by dgozman — COMMENTED** [review]
Two suggestions to fix failing tests, the rest is great!

**Review by dgozman — APPROVED** [review]
